ecs-agent/wsclient/client.go

```diff
@@ -0,0 +1,578 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package wsclient wraps the generated aws-sdk-go client to provide marshalling
+// and unmarshalling of data over a websocket connection in the format expected
+// by backend. It allows for bidirectional communication and acts as both a
+// client-and-server in terms of requests, but only as a client in terms of
+// connecting.
+package wsclient
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/cipher"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/httpproxy"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient/wsconn"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/private/protocol/json/jsonutil"
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+)
+
+const (
+	// ServiceName defines the service name for the agent. This is used to sign messages
+	// that are sent to the backend.
+	ServiceName = "ecs"
+
+	// wsConnectTimeout specifies the default connection timeout to the backend.
+	wsConnectTimeout = 30 * time.Second
+
+	// wsHandshakeTimeout specifies the default handshake timeout for the websocket client
+	wsHandshakeTimeout = wsConnectTimeout
+
+	// readBufSize is the size of the read buffer for the ws connection.
+	readBufSize = 4096
+
+	// writeBufSize is the size of the write buffer for the ws connection.
+	writeBufSize = 32768
+
+	// Default NO_PROXY env var IP addresses
+	defaultNoProxyIP = "169.254.169.254,169.254.170.2"
+
+	errClosed = "use of closed network connection"
+
+	// ExitTerminal indicates the agent run into error that's not recoverable
+	// no need to restart
+	ExitTerminal = 5
```

Comment:
Can we move the [exitcodes package](https://github.com/aws/amazon-ecs-agent/blob/883211e24d9255023a68f6351001c291af8a699b/agent/sighandlers/exitcodes/exitcodes.go) to /ecs-agent? Perhaps in a fast follow-up PR, if not this one.

Reply:
Sure. Will move it as a follow-up PR. This is already quite bulky. But thank you for pointing out.